### PR TITLE
Rubocop/standardrb

### PR DESCRIPTION
**Pull request by lukaszsliwa**
Resolves https://github.com/ably/ably-ruby/issues/139

### CODE REVIEW

**Comment by lukaszsliwa** [comment]
> @lukaszsliwa I see you're adding a custom `.rubocop.yml`, did you consider using a community maintained one, e.g. https://github.com/testdouble/standard?

@lmars 

I haven't checked stanradrb before but I see that standardrb the custom out rubocop.yml covers standardrb config.
This is working configuration in ably-ruby:

```ruby
parallel: true
format: progress
ruby_version: 2.7

ignore:
  - 'lib/submodules/**/*'
  - 'vendor/**/*'
```

And works fine:

```
200 files inspected, no offenses detected
```

`ruby_version` is set to 2.7 (the minimum supported version).

**Review by mattheworiordan — APPROVED** [review]
LGTM, tests passing too, so AFAICT a good set of changes 👍

File: lib/ably/models/channel_details.rb [hunk 1]
```diff
@@ -1,59 +1,65 @@
...
-
-    alias_method :to_h, :attributes
-
-    # Initialize a new ChannelDetails
+    # @param attributes (see #initialize)
     #
-    def initialize(attrs)
-      @attributes = IdiomaticRubyWrapper(attrs.clone)
+    # @return [ChannelDetails]
+    def self.ChannelDetails(attributes)
```
Comment (lawrence-forooghian) [inline lib/ably/models/channel_details.rb:12]: I haven't seen this `self.<module name>` syntax before – would you mind explaining what it does, and what is the violation that it’s replacing?

**Review by lawrence-forooghian — COMMENTED** [review]
I started reviewing this, but it's quite hard to review because it's a huge PR which doesn't require equal level of reviewer attention throughout. The changes here are (presumably) a mixture of fixes automatically generated by running `standard --fix`, and then some manual fixes that you've made for things that Standard couldn't automatically fix.

The automatic fixes don't really need to be reviewed (or at least need minimal reviewer attention), and the manual fixes need a good amount of reviewer attention. However, looking at the PR, there's no way for a reviewer to tell which is which.

When I run Standard, it says that there are `10605 offenses detected, 10546 offenses autocorrectable`. I think that this PR would be much more reviewable if there was one commit which runs `standard --fix`, and then one or more commits that manually address the 59 remaining issues.

File: lib/ably/models/channel_details.rb (hunk 1 above)
Comment (lukaszsliwa) [reply lib/ably/models/channel_details.rb:12]: @lawrence-forooghian 

The yardoc comment was added by me. The definition was not changed. I don't use this syntax but there are some more similar definitions in the gem.

Comment (lukaszsliwa) [reply lib/ably/models/channel_details.rb:12]: Automatic fix from `rubocop`, generated invalid code so most changes were made manually.
`standardrb --fix` were made in last commit https://github.com/ably/ably-ruby/pull/373/commits/a7971921cd1ddb2ab41894c29c1ec0bd04a0123a

File: lib/ably/auth.rb [hunk 2]
```diff
@@ -25,21 +27,21 @@ module Ably
...
+    include ::Ably::Modules::Conversions
+    include ::Ably::Modules::HttpHelpers
 
     # Default capability Hash object and TTL in seconds for issued tokens
     TOKEN_DEFAULTS = {
       renew_token_buffer: 10 # buffer to allow a token to be reissued before the token is considered expired (Ably::Models::TokenDetails::TOKEN_EXPIRY_BUFFER)
     }.freeze
 
-    API_KEY_REGEX = /^[\w-]{2,}\.[\w-]{2,}:[\w-]{2,}$/
+    API_KEY_REGEX = /^[\w-]{2,}\.[\w-]{2,}:[\w-]{2,}$/.freeze
```
Comment (TheSmartnik) [inline lib/ably/auth.rb:38]: Is is an offense? I'm pretty sure regex freeze is redundant
https://github.com/rubocop/rubocop/issues/6331

File: lib/ably/models/channel_details.rb (hunk 1 above)
Comment (TheSmartnik) [reply lib/ably/models/channel_details.rb:12]: @lawrence-forooghian it's a classic approach that is used when you want to convert to object to object of desired class unless it's already of the desired class. ruby uses it `Kernel::Float` `Kernel::Array`

File: lib/ably/models/device_push_details.rb [hunk 3]
```diff
@@ -1,86 +1,86 @@
...
 
-    %w(state).each do |attribute|
-      define_method attribute do
-        attributes[attribute.to_sym]
+        @raw_hash_object = hash_object || {}
+        @hash_object = IdiomaticRubyWrapper(@raw_hash_object)
       end
 
-      define_method "#{attribute}=" do |val|
-        unless val.nil? || val.kind_of?(String)
```
Comment (TheSmartnik) [inline lib/ably/models/device_push_details.rb:48]: Why this extra logic was removed?

File: lib/ably/models/message_encoders/base.rb [hunk 4]
```diff
@@ -93,37 +98,38 @@ def strip_current_encoding_part(message)
...
         # @return [Boolean]
         def is_empty?(message)
-          message[:data].nil? || message[:data] == ''
+          message[:data].nil? || message[:data] == ""
         end
       end
 
       # @api private
       def self.register_default_encoders(client, options = {})
-        binary_protocol = !!options[:binary_protocol]
```
Comment (TheSmartnik) [inline lib/ably/models/message_encoders/base.rb:102]: Up to you, but I think this rule is useless, only complicates code with extra condition

File: lib/ably/modules/enum.rb [hunk 5]
```diff
@@ -1,228 +1,221 @@
...
-          symbol
-        end
+          def to_sym
+            symbol
+          end
 
-        def to_json(*args)
-          %{"#{symbol}"}
-        end
+          def to_json(*_arg)
```
Comment (TheSmartnik) [inline lib/ably/modules/enum.rb:154]: `(def to_json(*_args)`

File: lib/ably/modules/model_common.rb [hunk 6]
```diff
@@ -1,77 +1,84 @@
...
+      end
 
-    # Stringify the JSON representation of this object from the underlying #attributes
-    # @return [String]
-    def to_json(*args)
-      as_json.to_json(*args)
-    end
+      # Return a JSON ready object from the underlying #attributes using Ably naming conventions for keys
+      # @return [Hash]
+      def as_json(*_arg)
```
Comment (TheSmartnik) [inline lib/ably/modules/model_common.rb:32]: `as_json(*_args)`?

**Review by TheSmartnik — COMMENTED** [review]
I've reviewed half of the files. A few minor comments
I'll review later this week

File: spec/unit/modules/conversions_spec.rb [hunk 7]
```diff
@@ -1,70 +1,72 @@
-require 'spec_helper'
+# frozen_string_literal: true
+
+require "spec_helper"
 
 describe Ably::Modules::Conversions, :api_private do
-  let(:class_with_module) { Class.new do; include Ably::Modules::Conversions; end }
+  let(:class_with_module) { Class.new { ; include Ably::Modules::Conversions; } }
```
Comment (TheSmartnik) [inline spec/unit/modules/conversions_spec.rb:6]: `let(:class_with_module) { Class.new { include Ably::Modules::Conversions }`

File: lib/ably/realtime/push.rb [hunk 8]
```diff
@@ -20,20 +22,21 @@ def admin
       # Activate this device for push notifications by registering with the push transport such as GCM/APNS
       #
       # @note This is unsupported in the Ruby library
-      def activate(*arg)
+      def activate(*_arg)
```
Comment (TheSmartnik) [inline lib/ably/realtime/push.rb:25]: `def activate(*_args)`

File: lib/ably/realtime/push.rb [hunk 9]
```diff
@@ -20,20 +22,21 @@ def admin
...
         raise_unsupported
       end
 
       # Deactivate this device for push notifications by removing the registration with the push transport such as GCM/APNS
       #
       # @note This is unsupported in the Ruby library
-      def deactivate(*arg)
+      def deactivate(*_arg)
```
Comment (TheSmartnik) [inline lib/ably/realtime/push.rb:32]: `def deactivate(*_args)`

File: lib/ably/rest/client.rb [hunk 10]
```diff
@@ -560,37 +553,37 @@ def raw_request(method, path, params = {}, options = {})
...
           conn = if preferred_fallback_connection_for_first_request
-            case retry_count
-            when 0
-              preferred_fallback_connection_for_first_request
-            when 1
-              # Ensure the root host is used first if the preferred fallback fails, see #RSC15f
-              connection(use_fallback: false)
-            end
-          end || connection(use_fallback: use_fallback) # default to normal connection selection process if not preferred connection set
+                   case retry_count
```
Comment (TheSmartnik) [inline lib/ably/rest/client.rb:569]: I'm strongly against this rule. It's extremely inconvenient. It's also inconsistent with other changes like removal of assignments alignment
```
foo      = :foo 
foobar = :foo
```

File: lib/ably/rest/middleware/exceptions.rb [hunk 11]
```diff
@@ -8,43 +10,38 @@ module Middleware
...
-                raise Ably::Exceptions::UnauthorizedRequest.new(*exception_args)
-              end
-            elsif env.status == 403
-              raise Ably::Exceptions::ForbiddenRequest.new(*exception_args)
-            elsif env.status == 404
-              raise Ably::Exceptions::ResourceMissing.new(*exception_args)
-            else
-              raise Ably::Exceptions::InvalidRequest.new(*exception_args)
-            end
+          return unless env.status >= 400
```
Comment (TheSmartnik) [inline lib/ably/rest/middleware/exceptions.rb:13]: `return if env.status < 400` reads better IMO

File: lib/ably/auth.rb (hunk 2 above)
Comment (lukaszsliwa) [reply lib/ably/auth.rb]: Fixed in 47f3bf7494f00fe6aaeaec2f3230f4f2737d3c3b

File: lib/ably/models/message_encoders/base.rb (hunk 4 above)
Comment (lukaszsliwa) [reply lib/ably/models/message_encoders/base.rb:102]: I was trying to fix as many offenses as possible. IMO, both `!!param` and `param ? true : false` are OK.

File: lib/ably/modules/enum.rb (hunk 5 above)
Comment (lukaszsliwa) [reply lib/ably/modules/enum.rb]: Fixed 771c3a7d4a3c6ac1f2073b0f50c95b5440fa0cd4

File: lib/ably/modules/model_common.rb (hunk 6 above)
Comment (lukaszsliwa) [reply lib/ably/modules/model_common.rb]: Fixed 771c3a7d4a3c6ac1f2073b0f50c95b5440fa0cd4

File: lib/ably/realtime/push.rb (hunk 8 above)
Comment (lukaszsliwa) [reply lib/ably/realtime/push.rb]: Fixed 771c3a7d4a3c6ac1f2073b0f50c95b5440fa0cd4

File: lib/ably/realtime/push.rb (hunk 9 above)
Comment (lukaszsliwa) [reply lib/ably/realtime/push.rb]: Fixed 771c3a7d4a3c6ac1f2073b0f50c95b5440fa0cd4

File: spec/unit/modules/conversions_spec.rb (hunk 7 above)
Comment (lukaszsliwa) [reply spec/unit/modules/conversions_spec.rb]: Fixed 9b41dac887f17ba109da0c7adc2f8f11962d5ec1

File: lib/ably/models/device_push_details.rb (hunk 3 above)
Comment (lukaszsliwa) [reply lib/ably/models/device_push_details.rb:48]: It wasn't removed. It's line 54 now.

File: lib/ably/rest/client.rb (hunk 10 above)
Comment (TheSmartnik) [reply lib/ably/rest/client.rb:569]: @lukaszsliwa can we please discuss this one

Comment (lukaszsliwa) [reply lib/ably/rest/client.rb:569]: @TheSmartnik What would you like to discuss?

**Comment by owenpearson** [comment]
Moving this temporarily to draft state as we have a lot of conflicts with the main branch that we need to fix. We plan to address some other higher priority work beforehand and will pick up this PR again in the near future.

**Pull request closed without merging**